torchrl/envs/libs/openspiel.py

```diff
@@ -0,0 +1,358 @@
+
+    def _check_kwargs(self, kwargs: Dict):
+        pyspiel = self.lib
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, pyspiel.State):
+            raise TypeError("env is not of type 'pyspiel.State'.")
+
+    def _build_env(self, env, requires_grad: bool = False, **kwargs):
+        self.requires_grad = requires_grad
+        return env
+
+    def _init_env(self):
+        self._update_action_mask()
+
+    def _make_specs(self, env: "pyspiel.State") -> None:  # noqa: F821
+        self.reward_spec = UnboundedContinuousTensorSpec(
+            shape=[*self.batch_size, env.num_players()],
+            device=self.device,
+        )
+        self.done_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size((1,)),
+            dtype=torch.bool,
+            device=self.device,
+        )
+        action_spec_cls = (
+            DiscreteTensorSpec
+            if self._categorical_action_encoding
+            else OneHotDiscreteTensorSpec
+        )
+        self.action_spec = action_spec_cls(
+            env.num_distinct_actions(),
+            dtype=torch.int64,
+            device=self.device,
+        )
+        self.observation_spec = CompositeSpec(
+            observation=UnboundedContinuousTensorSpec(
+                shape=(
+                    *self.batch_size,
+                    env.num_players(),
+                    *env.get_game().observation_tensor_shape(),
+                ),
+                device=self.device,
+            ),
+            current_player=DiscreteTensorSpec(
+                env.num_players(),
+                dtype=torch.int,
+                device=self.device,
+            ),
+            state=NonTensorSpec([]),
+            shape=self.batch_size,
+        )
+
+    def _set_seed(self, seed):
+        if seed is not None:
+            raise NotImplementedError("This environment has no seed.")
+
+    def current_player(self):
+        return self._env.current_player()
+
+    def _update_action_mask(self):
+        action_mask = torch.zeros(
+            self._env.num_distinct_actions(), device=self.device, dtype=torch.bool
+        )
+        action_mask[self._env.legal_actions()] = True
+        self.action_spec.update_mask(action_mask)
+
+    def _make_td_out(self, exclude_reward=False):
+        observation_shape = self._env.get_game().observation_tensor_shape()
+        observation = torch.stack(
+            [
+                self._to_tensor(self._env.observation_tensor(idx)).reshape(
+                    observation_shape
+                )
+                for idx in range(self._env.num_players())
+            ]
+        )
+        reward = self._to_tensor(self._env.returns())
+        done = torch.tensor(
+            self._env.is_terminal(), device=self.device, dtype=torch.bool
+        )
+        state = self._env.serialize()
```

Comment:
let's put a flag around that to make the state optional. Maybe "stateless=True"?

Reply:
What if we call the flag `output_state`? I think that name makes it more obvious what exactly it does--it adds the state to the output of `reset` and `step`

Reply:
I renamed to `return_state`, because that's what `PettingZooWrapper` calls it